### PR TITLE
Cp 19 Tiptap Editor Features

### PR DESCRIPTION
**Pull request by aatbip**
Task/Title
[Main page editor](https://outsidetech.atlassian.net/browse/CP-19?atlOrigin=eyJpIjoiM2JkYTI2M2U3NWJhNDRiOGI3N2Q1ZTA4OTdmNjRkM2UiLCJwIjoiaiJ9)

#### The main changes are

- [x] tiptap library setup, 
- [x] added floating menu component's layout
- [x] floating menu: heading, list, paragraph
- [x] floating menu: table list  
- [x] bubble menu: dropdown text formatters, bold, underline, strike, italic
- [x] floating menu: link, callout, embed

### CODE REVIEW

File: src/app/components/ClientHomeInterface.tsx [hunk 1]
```diff
@@ -15,10 +16,18 @@ const ClientHomeInterface = () => {
 
   const [selected, setSelected] = useState(defaultValue)
```
Comment (saugatmaharjan) [inline src/app/components/ClientHomeInterface.tsx:17]: Let's not use generic name for state, use name that defines/explains the actual functionality/usage

File: src/app/components/EditorInterface.tsx [hunk 2]
```diff
@@ -2,27 +2,123 @@
...
+import CalloutExtension from '@/components/tiptap/callout/CalloutExtension'
+import Bold from '@tiptap/extension-bold'
+import Italic from '@tiptap/extension-italic'
+import Strike from '@tiptap/extension-strike'
+import Gapcursor from '@tiptap/extension-gapcursor'
+
+import LinkInput from '@/components/tiptap/linkInput/LinkInput'
+import FloatingMenuContainer from '@/components/tiptap/floatingMenu/FloatingMenu'
+import BubbleMenuContainer from '@/components/tiptap/bubbleMenu/BubbleMenu'
+import { useEffect } from 'react'
```
Comment (saugatmaharjan) [inline src/app/components/EditorInterface.tsx:32]: import order

File: src/app/components/EditorInterface.tsx [hunk 3]
```diff
@@ -2,27 +2,123 @@
...
+import AutofieldSelector from '@/components/tiptap/autofieldSelector/AutofieldSelector'
 
 const EditorInterface = () => {
   const appState = useAppState()
 
+  const editor = useEditor({
+    extensions: [
```
Comment (saugatmaharjan) [inline src/app/components/EditorInterface.tsx:39]: may be we can define this configuration outside of the component and memoize or use useRef

File: src/app/components/EditorInterface.tsx [hunk 4]
```diff
@@ -2,27 +2,123 @@
...
         style={{
-          background: '#f8f9fb',
-          height: '90vh',
+          background: '#f8f9fb', //to be changed later with color picker component
+          maxHeight: '91vh',
         }}
       >
-        EDITOR
+        {
+          editor ? (
```
Comment (saugatmaharjan) [inline src/app/components/EditorInterface.tsx:112]: do we need this this check? component already return null if editor is not defined

File: src/components/autofillFields/AutofillFields.tsx [hunk 5]
```diff
@@ -0,0 +1,18 @@
+
+const AutofillFields = () => {
+  return (
+    <div className="p-5">
+      <p className="font-medium pb-5">Autofill fields</p>
+
+      <div className="flex flex-col gap-5">
+        <p className="text-new-gray">	&#123;&#123;client.givenName&#125;&#125;</p>
```
Comment (saugatmaharjan) [inline src/components/autofillFields/AutofillFields.tsx]: let's create separate component to render auto field

File: src/components/tiptap/autofieldSelector/AutofieldSelector.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+        }} pluginKey="autofield-selector">
+
+        <div className='flex flex-col gap-0.5 bg-white py-2 border border-new-card-border rounded shadow-vairant-1 absolute top-3 w-52'>
+          <button className='flex flex-row gap-x-2.5 items-center py-1.5 px-3 focus:bg-new-white-2 cursor-pointer outline-none'>	&#123;&#123;client.givenName&#125;&#125;</button>
+          <button className='flex flex-row gap-x-2.5 items-center py-1.5 px-3 focus:bg-new-white-2 cursor-pointer outline-none'>	&#123;&#123;client.familyName&#125;&#125;</button>
+          <button className='flex flex-row gap-x-2.5 items-center py-1.5 px-3 focus:bg-new-white-2 cursor-pointer outline-none'>	&#123;&#123;client.email&#125;&#125;</button>
+          <button className='flex flex-row gap-x-2.5 items-center py-1.5 px-3 focus:bg-new-white-2 cursor-pointer outline-none'>	&#123;&#123;client.company&#125;&#125;</button>
+          <button className='flex flex-row gap-x-2.5 items-center py-1.5 px-3 focus:bg-new-white-2 cursor-pointer outline-none'>	&#123;&#123;client.address&#125;&#125;</button>
+        </div>
```
Comment (saugatmaharjan) [inline src/components/tiptap/autofieldSelector/AutofieldSelector.tsx]: separate component

File: src/components/tiptap/callout/Callout.tsx [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { NodeViewContent, NodeViewWrapper } from '@tiptap/react'
+import React from 'react'
+
+export const Callout = () => {
+  return (
+    <NodeViewWrapper className="callout-container">
+      {/* <span className="label" contentEditable={false}>Callout</span> */}
```
Comment (saugatmaharjan) [inline src/components/tiptap/callout/Callout.tsx:7]: let's remove commented code if not necessary

File: src/components/tiptap/bubbleMenu/BubbleMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+              }}
+            >
+              <div>
+                <H1Icon />
+              </div>
+              <div>
+                <p className='text-sm'>Heading 1</p>
+              </div>
+            </button>
+          </MenuItem>
```
Comment (saugatmaharjan) [inline src/components/tiptap/bubbleMenu/BubbleMenu.tsx]: let's separate this in to another component

File: src/components/tiptap/floatingMenu/FloatingMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+              tiptapEditorUtils.toggleHeading(2)
+            }}
+          >
+            <div>
+              <H2Icon />
+            </div>
+            <div>
+              <p className='text-sm'>Heading 2</p>
+            </div>
+          </button>
```
Comment (saugatmaharjan) [inline src/components/tiptap/floatingMenu/FloatingMenu.tsx]: whenever there is some repetitive code, try extract that to separate component, to make code clean and maintainable

File: src/context/index.tsx [hunk 10]
```diff
@@ -18,19 +22,33 @@ interface IAppCoreProvider {
...
+    bannerImg: "",
+    showLinkInput: false,
+    readOnly: false
   })
 
   const setBannerImg = (image: string) => {
     setState((prev) => ({ ...prev, bannerImg: image }))
   }
 
+  const toggleShowLinkInput = (v: boolean) => {
```
Comment (saugatmaharjan) [inline src/context/index.tsx:35]: avoid using acronym

**Comment by sazanrjb** [comment]
please resolve conflicts @aatbip

**Comment by aatbip** [comment]
@saugatmaharjan I will fix the above issues in another PR.

**Review by sazanrjb — APPROVED** [review]

**Review by saugatmaharjan — APPROVED** [review]

**Pull request merged**